### DISCUSSION TITLE
Chapter Editor redesign

### DISCUSSION THREAD

**Discussion by nichwall** [advplyr/audiobookshelf#3011] [Ideas]
In trying to figure out the best way to show the Lookup chapters next to the existing chapters (mentioned in https://github.com/advplyr/audiobookshelf/issues/2640), I spent some time trying to redesign the Chapter page. Afterwards, I found some similar requests which touch on some of the same points.
- https://github.com/advplyr/audiobookshelf/issues/2816
- https://github.com/advplyr/audiobookshelf/issues/2225

The main things to change in the layout:
- Move the book duration to just below the title.
- Keep the "use seconds instead of timestamps" checkbox, but also add "use duration instead of start time" and have those apply to *all* of the time related displays.
- Make sure the titles of the left and right halves are aligned.
- Left half:
  - In the left half, make the "Remove All" instead just clear the left side and not remove all chapters from the library item. This way users can still "Revert" or do some more manual edits before the chapters are deleted from the item.
  - Redesign the "shift times" and not have it be collapsible so both sides stay in alignment.
    - This could be by having a text field and a plus/minus on either side to automatically apply the seconds difference. Or, having 4 buttons, `-6` `-1` `+1` `+6`. 6 was chosen because that is the length of the Audible intro.
  - Move the "add, remove, play" chapter buttons to the left? Not really sure where to put those.
  - The timestamp of the very first chapter is greyed out and cannot be changed from `00:00:00`. This makes it visually distinct and helps users to see that the start time will not change by using the "shift times"
  - It might be nice if the chapters are draggable to more easily reorder, but that would make the "start times" a bit annoying to clean up.
- Right half:
  - Make the right side have 3 tabs: `Tracks`, `ID3 Chapters`, `Lookup` and make the rows align with the left side so it's easier to visually compare them. The right side is not editable and is used for information purposes only. Each one of the 3 tabs has a "set chapters" button that removes all chapters from the left and copies data from the right to the left like it currently does for filenames.
    - `Tracks`: remove the "Chapters" checkbox column and move the "time" column to the left of the filename so all 3 tabs have the same layout. By default the time column is "start time".
    - `ID3 Chapters`: add a "probe audio files" button that gets the chapters and builds the list on the right side. If no embedded tracks are found, then display a message.
    - `Lookup`: move the current Audible chapter lookup out of the modal and put it in this main tab. Have the ASIN, Region, and search buttons in the top so it is easier to search again with different criteria.
- Add a "copy arrow" between the columns on the left and right to make updating individual chapters easier. This arrow would not appear on narrow screens because the "right" becomes the "top"
- It would be nice if chapters in the left side could be draggable like in the `Track Editor`, but not sure how hard that would be to add.

I spent some time trying to figure out how to make this intuitive on a mobile device or smaller screen, and my conclusion is "this should really be done with an actual monitor, but the new layout does not make the mobile experience worse." This should also be manageable with screen readers, but not sure how well the "copy arrow" functionality would work for that.

![PXL_20240525_182737575](https://github.com/advplyr/audiobookshelf/assets/5686638/c0162bb7-6c20-4cad-98e2-426ab347b47e)

<details>
  <summary>Narrow Screen behavior</summary>

  ![image](https://github.com/advplyr/audiobookshelf/assets/5686638/15178f12-38d0-4160-a1f9-9af9cf8e08a6)
</details>

**Comment by advplyr** [comment]
I like the idea of having tabs and showing chapters side-by-side for comparison.

I'm not sure how dragging would work with regard to the start time. I think the start times should stay ascending. I only had one case just yesterday where I wanted to drag a chapter and it was due to a bug that I just fixed. Other than that I can't think of why I'm dragging.

I like the option to re-probe audio files for embedded chapters. I have to think about this tab some more because one thing that was useful was being able to tell quickly if a book had multiple files with embedded chapters or one when debugging issues. If there are multiple files with embedded chapters they might be merged.

For #2816 I think we would have to show this on the tracks tab. A toggle between filename and ID3 title if it is present or something like that.

I can't think of a scenario where the arrow between each chapter row would be useful. Each tab would need similar buttons that are now on the lookup modal, "Apply Chapters" and "Map Chapters".

One thing I've been wanting to update on the chapter editor probably the most of all is the sample audio feature. I think when sampling the chapter this needs to show a separate audio player somewhere that allows seeking and showing the timestamp. I haven't thought of a good way to implement that yet but I use the sample feature the most when checking if times match up to the audio.

**Reply by nichwall** [reply]
Yeah, dragging doesn't really fit. The main use case I was thinking of is if you want to drag say "chapter 5" from the right to be after "chapter 7" because there are extra introduction chapters, but adding a row then copy/pasting should be good enough for that. This logic also applies to the arrows so they could be removed.

A toggle between the tracks and ID3 tabs would work. I was thinking that separating them out a bit would make it easier to see where the chapters on the left came from. Maybe having something like the file name be small above the first chapter that is in the track. This would be difficult for chapters that span multiple tracks.

<details>
  <summary> File names above ID3 chapter titles </summary>

![PXL_20240525_205637279](https://github.com/advplyr/audiobookshelf/assets/5686638/76fc8780-defd-4511-bfe4-6e2a10c8094d)
</details>

Yes, each tab would need the "Apply Chapters" and "Map Chapters" button.

Maybe it could just be a little player box that appears at the top of the left column? All it would really need is the "play/pause" button, a "forward/backward" button, and a timestamp display. It could show a track bar like in the normal web player, but be zoomed it to 5 minutes in both directions or something instead of showing the entire book.
![image](https://github.com/advplyr/audiobookshelf/assets/5686638/231e115a-480a-46da-9f1e-d7e0d253ba37)

**Reply by advplyr** [reply]
> A toggle between the tracks and ID3 tabs would work. I was thinking that separating them out a bit would make it easier to see where the chapters on the left came from. Maybe having something like the file name be small above the first chapter that is in the track. This would be difficult for chapters that span multiple tracks.

I misunderstood what you originally meant by the ID3 Chapters tab. I was thinking this would be the tab that shows the embedded chapters and allow for re-probing that info.
The 2 ways chapters are set are either using the tracks or using embedded chapter markers. When using tracks for chapters there is a check to see if the ID3 title tags meet a criteria.
https://github.com/advplyr/audiobookshelf/blob/6fa49e0aab2309d6f64d08dcc18343f6d4c0694d/server/scanner/AudioFileScanner.js#L492-L498

**Reply by nichwall** [reply]
Oh, yes I was thinking the `ID3 Chapters` tab would probe that info like you said with the embedded chapter markers. I was thinking the user can still have the option for which ones to apply in ABS and just include them all in this tab. I don't really use embedded chapters so I am probably mixing things up with them.

**Reply by advplyr** [reply]
Okay yeah the ID3 chapter titles are just for when using tracks to make chapters so that is why I was thinking they should be shown in the tracks tab somehow.

**Reply by advplyr** [reply]
The `ID3 Chapters` tab would be named `Embedded Chapters` or something like that, since it's not ID3. Although it could be ID3 if they have Overdrive Chapter markers set.

**Comment by ZLoth** [comment]
First off, I love the book "Match" feature that pulls the data from a data source (usually Audible) and updates the information displayed in both the Details and Cover tab. I think the same principal should apply to editing chapters.

The challenge is that while Audible does a good job of dividing a book into time-stamped chapters, a significant number of titles have just the numbered chapters without actually adding the chapter labels including Prologue, Epilogue, Epigraph, and what have you. I end up taking the converted MP3 files and adding my own ID3 labels (embedded chapters) for each chapter after doing research (looking up the Kindle version of the book and pulling those chapter titles). So, that gives up several possibilities of importing the chapter data:

* Metadata from Amazon (if ASIN is available)
* Filenames (which is present)
* Embedded chapters (ID3 tags)
* Export/import as Comma-separated or tab-separated file (possible future functionality)

The left-most tab would be the existing chapter editor. The other tabs are import functions so that a administrator can preview the timestamp and chapter title, then click the "Set chapters from tracks" to update the Chapter Editor. That data would not be saved until the "Save" button is clicked in the chapter editor. 

Having it as tabs would provide possible future import options while keeping the interface clean. 

Here are some rough visualizations based upon the existing design:
![image](https://github.com/advplyr/audiobookshelf/assets/6700159/2c3c1c81-1733-4e70-9b78-8225a6e1392c)

![image](https://github.com/advplyr/audiobookshelf/assets/6700159/d4521887-2a8e-4504-a825-58790a0d420b)
Metadata and Embedded Chapters would follow the same design.

**Comment by ZLoth** [comment]
Following up on the status of this issue.

**Reply by nichwall** [reply]
I haven't worked on this at all since the initial discussion

**Reply by ZLoth** [reply]
Acknowledged. This is something that I really need in organizing my book collection. :D

**Comment by ZLoth** [comment]
Following up on the status of this feature request. Yes, I'm very much interested in the import from metadata function to clean up the chapters in my library.

**Comment by iconoclasthero** [comment]
I guess this is a good place to add some ideas...
I was just trying to align a CD rip to the chapter markers from Audible for a 38-hour book biography of a famous Georgian (so there were a lot of foreign names making it more difficult to find things in the eBook! :).  

Two things would make the aligning easier:

1. A +/- adjustment on the left for the time rather than dealing with having to change individual numerals in the display.  (Also, a local undo for an accidentally-deleted chapter if you think the +/- on the right is for time!)
2. A clock display of the time elapsed from the start of the chapter when you hit the little play button...or, at the very least, some way to copy the current timestamp to the chapter start time field such that when I hear "Chapter 14" it copies _that_ time to the timestamp field and then I can manually adjust down a second or two from there.

Pursuant to 2, a clock would help the editor determine the actual delay between when  ▶  is clicked and when sound starts coming out, allowing for differentiation between dead air and buffering.

An actual player in which there was a clock and a way to adjust the playback speed would also be very helpful.  ABS doesn't actually need a playback speed controller if I can use the chrome plugin I have for other things.

More blue sky, it would be nice if there was a way to either progressively shift the time stamps (e.g., as the CD version gets longer and longer from 00:00:00.00 origin vs. an eAudiobook, a progressive amount could be applied.  This seems somewhat difficult to implement so a more reasonable way to do it would be to allow the user to select _where in the book_ to start the application of time shifts.  Instead of it applying to the whole book, one could adjust the whole thing back 3 seconds and then move to the middle point of the book and apply another -2 seconds, etc.  This isn't wholly unrelated: https://github.com/advplyr/audiobookshelf/issues/1533

Another blue-sky idea would be to allow for users to convert bookmarks made in the player to chapters.

NB: I would still like to be able to see a two-column list comparing the existing chapter index to whatever's been pulled from Audible to see if they're even remotely close...  This would be in the chapter lookup dialog, not in the editor per se.

**Reply by iconoclasthero** [reply]
> More blue sky, it would be nice if there was a way to either progressively shift the time stamps (e.g., as the CD version gets longer and longer from 00:00:00.00 origin vs. an eAudiobook, a progressive amount could be applied. This seems somewhat difficult to implement so a more reasonable way to do it would be to allow the user to select _where in the book_ to start the application of time shifts. Instead of it applying to the whole book, one could adjust the whole thing back 3 seconds and then move to the middle point of the book and apply another -2 seconds, etc. This isn't wholly unrelated: #1533

A straightforward way to apply progressive shifts as the chapters drift more or less is this:

Add a `Lock [time]` checkbox to the left of the timestamp box or between it and the title.  There would also be a global select/deselect all box at the top.

So on `Moscow 1812` the workflow would have been:

Import the timestamps without the audible intro and outro.  Add 4 seconds to the entire book.
- Entry 1 doesn't change: lock doesn't matter.
- Entry 2 manually adjusted -1 second and locked
- After auditorially checking entries 3-12, they're all locked
- Entries 13-28 are adjusted -1 second
- Save, done.

---
If a book is longer and drifts more as the book gets further from the start, more chapters can be locked and additional time can be added to the later chapters as one goes along.  

E.g., for a 100-chapter book, chapters 1-10 are acceptable and locked; chapters 11-100 get an additional second of adjustment.  Chapters 11-29 are now acceptable and locked, but there's more drift at chapter 30 so an additional adjustment second is added for 30-100, etc., etc., until the whole book's aligned properly.  This gets around the necessity of having to individually add seconds to chapters individually and should vastly speed up the process.

***
**I think this is something that should be relatively easy to implement and doesn't necessarily require a whole redesign.**

**Comment by DarkHeraldMage** [comment]
What I'd really like to see is an option to adjust the time starting at a specific number in the list. For instance, I can shift everything +2 seconds if needed, but I often find that 20 files in (a random example, it's not always that number) I need to shift another second, and then 30 more files and I'm having to shift another second again. It's not exact for the whole book, especially long books, and since we can't do anything more granular than a second it can get tedious.

Right now you can only shift for every file all at once, but as I go through them I'd like to say to add or subtract time starting at a specific chapter marker, ignoring all the ones already reviewed. Or even better, have an option to flag which have been reviewed or confirmed as accurate; right now on my longer books I often stop in the middle so I just update the last chapter I confirmed to have something like "===========START HERE============" in the name so I can quickly find it in the list when I come back.

**Reply by iconoclasthero** [reply]
like this? 
https://github.com/advplyr/audiobookshelf/discussions/3011#discussioncomment-13272663

**Comment by iconoclasthero** [comment]
As I've been working a lot with this lately, I'd really like to be able to add the power of tools outside of the web interface to be able to edit chapters more efficiently.

I was thinking that more "import" options would be helpful...  
(I think these ideas are additional to https://github.com/advplyr/audiobookshelf/discussions/3011#discussioncomment-9558911.)
Where there now is `Audio Tracks — Set chapters from tracks (i)` there could be an "Advanced" or "Import" button.

1. **Import from embedded chapter index:** This would allow the user to import not just from the file names of books with tracks as files, but it would also allow the user to import chapters from an indexed opus file.  (Importing from m4b would be fine too, but not useful for me.)  If you have a file that already has a perfectly cromulent chapter index embedded in the audio file, why not be able to leverage that in setting chapter titles and times?

2. **Import from json:** This could be the metadata.json or it could be a separate, user-selectable json with the same chapter structure as metadata.json.  This allows the user to leverage cli tools like jq to do all kinds of manipulation and then import the resulting .json metadata smoothly and efficiently without crazy workarounds to get ABS to read the structured json data.  Since ABS already uses the json data structure, this seems much more straight forward than a csv or even the syntax that ffmpeg uses, though I see nothing prohibitive of using those as options as well.

```
  "chapters": [
    {
      "id": 0,
      "title": "Herbie Hancock -- Opening",
      "start": 0,
      "end": 31
    },
    {
      "id": 1,
      "title": "Chapter 01",
      "start": 31,
      "end": 1626.557
    },
   ⋮
       {
      "id": 23,
      "title": "Chapter 23",
      "start": 41469.817,
      "end": 43659.505
    },
    {
      "id": 24,
      "title": "Chapter 24",
      "start": 43659.505,
      "end": 45227.215625
    }
  ],
```

3. **Free-text parser:** Another possibility would be a paste-in parser that would allow the user to copy the existing index, edit it in another editor, i.e., something with find/replace function and then paste in updated chapter information.  This might get around issues of security with importing files.

**Reply by iconoclasthero** [reply]
A button somewhere in the edit (e.g., match > from metadata.json) dialogues to **selectively** import data from the metadata.json file would be simple and effective.

**Comment by iconoclasthero** [comment]
Also in the edit chapters (or edit > tools > embed metadata manager) dialog, a .cue export.  See https://github.com/advplyr/audiobookshelf/issues/909.

**Comment by iconoclasthero** [comment]
- **Allow for more than 59 seconds when adjusting the chapter start times.**  This would make fine, manual adjustments easier as you would not have to modify so many digits when the book is close to the end of a minute.  This obviously needs to change the minutes/hours commensurately either when the index is saved to DB & metadata.json, though I think that it would be more helpful if it recalculated hh:mm:ss based on (( ss > 59 )) when the play button was hit.

- **Display more bibliographical information** in the chapter editor so one can find the book on disk (command line/file browser): at a minimum, the author and the subtitle.  A path would also be helpful.
- **Always display book in home after closing chapter editor:** After opening the book from the "Discover" section of the "Home" view and going into the chapter editor, when it is closed, the book's gone.  It would be nice if it wasn't.

**Comment by chyron8472** [comment]
I would like the ability to bulk-edit chapters, and/or to be able to import chapter titles into the editor from a file.

**Reply by nexxai** [reply]
I'm not sure if what I'm about to suggest falls in line with this suggestion already, but it would be really nice to be able to bulk-rename chapters.  

I primarily listen to books in my car and the app I use to listen to them always adds the total remaining time in the book after the chapter name on the car's entertainment unit.  There's not many characters available in that space so I typically end up having to rename the chapters from "Chapter 1" to "001" or whatever.

It would be amazing if there was a way to submit some kind of formatter (hell, I don't even care if it's regex) that would apply to all chapter names.  

As it stands right now, what I end up doing is modifying a single chapter name, watching my browser's traffic, copying as cURL the POST to update that one chapter, do a find and replace on "Chapter", and then submit the updated cURL command.

**Reply by iconoclasthero** [reply]
> I'm not sure if what I'm about to suggest falls in line with this suggestion already, but it would be really nice to be able to bulk-rename chapters.
> 
> I primarily listen to books in my car and the app I use to listen to them always adds the total remaining time in the book after the chapter name on the car's entertainment unit. There's not many characters available in that space so I typically end up having to rename the chapters from "Chapter 1" to "001" or whatever.
> 
> It would be amazing if there was a way to submit some kind of formatter (hell, I don't even care if it's regex) that would apply to all chapter names.
> 
> As it stands right now, what I end up doing is modifying a single chapter name, watching my browser's traffic, copying as cURL the POST to update that one chapter, do a find and replace on "Chapter", and then submit the updated cURL command.

You should be able to do this with a shell script.  I've not looked at mine in a while, but it does stuff like this...

**Reply by nexxai** [reply]
> You should be able to do this with a shell script. I've not looked at mine in a while, but it does stuff like this...

Yeah, what I'm doing is *basically* that; was really just hoping for a more integrated solution than manually working a script.

**Reply by iconoclasthero** [reply]
> > You should be able to do this with a shell script. I've not looked at mine in a while, but it does stuff like this...
> 
> Yeah, what I'm doing is _basically_ that; was really just hoping for a more integrated solution than manually working a script.

Here's my function that does most of the work if you want to hack around it: https://github.com/iconoclasthero/indexopus/blob/2f8f77f4ce5511ccedec5b0782e223c0db9828b1/indexopus-3#L249

You can hit me up on Discord if you want to discuss, but the time frame for any further updates to the chapter interface is going to be lengthy unless you want to do it yourself.  The1ne did a great job on what's there now.

The script could be automated with systemd timers or inotifywait...

**Comment by chyron8472** [comment]
Also, I would appreciate being able to specify a start time and an end time to the book that isn't the actual beginning and end. Then I could omit the "This is Audible" bookends even if they hadn't been removed from the file. Currently, ABS insists that the first chapter starts at 00:00:00.000

**Comment by JustTooKrul** [comment]
I filed my own enhancement just now (https://github.com/advplyr/audiobookshelf/issues/4406) and this got flagged, so I figured I'd add my thoughts. 

First, I agree with the above that multiple importing methods--especially from a text file--is easiest and gives the user a lot of power. But, I think the most powerful method for importing chapters probably should take some cues from how cover art is dealt with. I ran into multiple books where the chapters had names (and you could look them up in the ebook), but all the newer editions had chapter titles of "Chapter [x]" without the names. I had to go back to older editions, get the ASNs and then use those ASNs, one-by-one, in the "Lookup" dialog. Once I found an edition with chapter names, I then had to edit the chapters in the book to match (since those would often have an extra chapter for "Introduction" or "Part 2" where the more recent editions just had the chapters matching the book's chapters) so when I imported it I could then make a few tweaks and have something where the chapters aligned (if you try to map an imported chapter section into an audio file with fewer chapters then you just get everything shifted, and shifting back is a chapter-by-chapter process). 

When importing covers, by contrast, you can search by title and see all your options--something similar for chapters would be ideal. And then having the ability to manually select which of the found chapters you want to import (currently it's "all or none") would be exceptionally helpful since you could map the imported chapters to your audio file(s) without having to manually edit the chapters in the audiobook first (I was primarily dealing with m4a files, converted into that format if necessary--for individual MP3s it would obviously be different).

But, as mentioned above, the simplest solution is to be able to import a text file and just rename things that way--it's simple, powerful, and probably has lowest overhead in terms of UI.

**Comment by iconoclasthero** [comment]
This applies specifically to https://github.com/advplyr/audiobookshelf/pull/4384 which I've been testing out for a week or so now (and assumes that it will be implemented)...

In using PR-4384, I would find value in a shift times dialog at the bottom of the list in addition to the one at the top so I don't have to keep scrolling up and down as I progress through the book and/or a hotkey with a pop-up dialog so I don't have to scroll anywhere.
Suggestions for hotkeys (i don't care if it's ^ or ALT or ^-ALT):
- ^S for shift
- ^A for align
- ^T for shift time

I assume that a modal would be easier to implement, but a modeless dialog that could be moved as one goes down the chapter list might be more useful.

In discussing it on Discord (DMs) The1ne responded: 
> I don't think I'll add a hotkey for that. If you want to try and add it that's fine. I don't have the motivation to add more to that PR right now or the brain power to think through how I would want to implement it and display the changes. Sorry man

So I'm putting it here to keep track of it for future development.

**Comment by iconoclasthero** [comment]
Separate buttons for [SAVE] and [SAVE & CLOSE] would be much appreciated.  I want to be able to not lose my updates as I go along and fix chapters without having to navigate back to the chapter edit function....and relock all the chapters I just fixed.

And a way to turn off tool tips...  they get, annoyingly, get in the way a lot.
